src/Backups/BackupCoordinationRemote.cpp

```diff
@@ -430,72 +483,92 @@ void BackupCoordinationRemote::prepareReplicatedAccess() const
-                replicated_access->addFilePath(access_zk_path, type, host_id, file_path);
+                String access_zk_path = unescapeForFileName(escaped_access_zk_path);
+                String path2 = path + "/" + escaped_access_zk_path;
+                for (const String & type_str : zk->getChildren(path2))
+                {
+                    AccessEntityType type = AccessEntityTypeInfo::parseType(type_str);
+                    String path3 = path2 + "/" + type_str;
+                    for (const String & host_id : zk->getChildren(path3))
+                    {
+                        String file_path = zk->get(path3 + "/" + host_id);
+                        replicated_access->addFilePath(access_zk_path, type, host_id, file_path);
+                    }
+                }
             }
-        }
-    }
+        });
 }
 
-
 void BackupCoordinationRemote::addFileInfo(const FileInfo & file_info, bool & is_data_file_required)
 {
-    auto zk = getZooKeeper();
+    auto zookeeper_retries_info = global_zookeeper_retries_info;
+    ZooKeeperRetriesControl retries_ctl("addFileInfo", zookeeper_retries_info);
+    retries_ctl.retryLoop(
+        [&]()
+        {
+            auto zk = getZooKeeper();
 
-    String full_path = zookeeper_path + "/file_names/" + escapeForFileName(file_info.file_name);
-    String size_and_checksum = serializeSizeAndChecksum(std::pair{file_info.size, file_info.checksum});
-    zk->create(full_path, size_and_checksum, zkutil::CreateMode::Persistent);
+            String full_path = zookeeper_path + "/file_names/" + escapeForFileName(file_info.file_name);
+            String size_and_checksum = serializeSizeAndChecksum(std::pair{file_info.size, file_info.checksum});
+            zk->createIfNotExists(full_path, size_and_checksum);
 
-    if (!file_info.size)
-    {
-        is_data_file_required = false;
-        return;
-    }
+            if (!file_info.size)
+            {
+                is_data_file_required = false;
+                return;
+            }
 
-    full_path = zookeeper_path + "/file_infos/" + size_and_checksum;
-    auto code = zk->tryCreate(full_path, serializeFileInfo(file_info), zkutil::CreateMode::Persistent);
-    if ((code != Coordination::Error::ZOK) && (code != Coordination::Error::ZNODEEXISTS))
-        throw zkutil::KeeperException(code, full_path);
+            full_path = zookeeper_path + "/file_infos/" + size_and_checksum;
+            auto code = zk->tryCreate(full_path, serializeFileInfo(file_info), zkutil::CreateMode::Persistent);
+            if ((code != Coordination::Error::ZOK) && (code != Coordination::Error::ZNODEEXISTS))
+                throw zkutil::KeeperException(code, full_path);
 
-    is_data_file_required = (code == Coordination::Error::ZOK) && (file_info.size > file_info.base_size);
+            is_data_file_required = (code == Coordination::Error::ZOK) && (file_info.size > file_info.base_size);
```

Comment:
Maybe a stupid suggestion - but what if we will write our hostname in that node in addition to the serialized file info? Then we will be able to distinguish between these situations, but we will waste a lot of memory in Keeper to store it in each node.

Reply:
Then we will not be able to distinguish:
 - It was normally created before (just like it could be without retries)
 - We lost the connection after the request was processed by ZooKeeeper and we tried to create it again

We need to carefully read the code around and find out who, when, and why creates that node (could it be created by the same replica?)

Reply:
> what if we will write our hostname in that node in addition to the serialized file info? Then we will be able to distinguish between these situations, but we will waste a lot of memory in Keeper to store it in each node.

Yes, I think this is how that can be implemented. Storing host names will waste a lot of memory, that's right, but you don't have to store the full host names (which can be long), it's enough to store just host indices (they're constant in the context of the backup operation). I made my PR https://github.com/ClickHouse/ClickHouse/pull/47971 to make using host indices easier